src/VisualStudio/Core/Def/Implementation/LanguageClient/AbstractInProcLanguageClient.cs

```diff
@@ -81,6 +85,7 @@ internal abstract class AbstractInProcLanguageClient : ILanguageClient
             _listenerProvider = listenerProvider;
             _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
             _diagnosticsClientName = diagnosticsClientName;
+            _asyncServiceProvider = (VSShell.IAsyncServiceProvider)asyncServiceProvider;
```

Comment:
I don't recall, @jasonmalinowski does this cast require the UI thread?

Reply:
So minimally we should have this cast up at the derived code -- otherwise we're separating the cast from the actual MEF import, and that's assuming an implementation detail more directly than we may want. Put another way, if this constructor is going to assume it really implements a different interface, write that in the signature to start with.

So assuming you are moving the cast to the importing constructor, I _believe_ this is fine because SVsServiceProvider in the MEF container is effectively managed so this cast is safe, but I think at this point you can also import IAsyncServiceProvider or something else directly to avoid the cast. We still have some code bouncing around Roslyn that does this cast because at the time there wasn't something better.

@AArnott I think you'll know what the current recommended way is to import an async service provider? I see https://github.com/microsoft/vs-mef/issues/138 is still active which makes me wonder what the current recommendation is.

Reply:
The SVsServiceProvider object should not be assumed to implement `IAsyncServiceProvider`, I'm pretty sure. You should query for `SAsyncServiceProvider` instead and cast that to `IAsyncServiceProvider`. Since that query is on an `IServiceProvider`, yes you need to be on the main thread to do it.
Alternatively though, if this is a MEF importing constructor you should be able to import `SAsyncServiceProvider` directly, as explained on that vs-mef issue that @jasonmalinowski already found.
Does that help?

Reply:
> You should query for SAsyncServiceProvider instead and cast that to IAsyncServiceProvider

Alright, we'll switch over to that generally. Thanks for confirming.

Reply:
Switched to SAsyncServiceProvider.